server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -0,0 +1,543 @@
+            if (providedValue == null) {
+                throw new IllegalArgumentException("\"" + column + "\" must not be null");
+            }
+        }
+    }
+
+    record CheckGeneratedValue(Input<?> input, GeneratedReference ref) implements ColumnConstraint {
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void verify(Object providedValue) {
+            DataType<Object> valueType = (DataType<Object>) ref.valueType();
+            Object generatedValue = input.value();
+            int compare = Comparator
+                .nullsFirst(valueType)
+                .compare(
+                    valueType.sanitizeValue(generatedValue),
+                    valueType.sanitizeValue(providedValue)
+                );
+            if (compare != 0) {
+                throw new IllegalArgumentException(
+                    "Given value " + providedValue +
+                    " for generated column " + ref.column() +
+                    " does not match calculation " + ref.formattedGeneratedExpression() + " = " +
+                    generatedValue
+                );
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public Indexer(String indexName,
+                   DocTableInfo table,
+                   TransactionContext txnCtx,
+                   NodeContext nodeCtx,
+                   Function<ColumnIdent, FieldType> getFieldType,
+                   List<Reference> targetColumns,
+                   Symbol[] returnValues) {
+        this.symbolEval = new SymbolEvaluator(txnCtx, nodeCtx, SubQueryResults.EMPTY);
+        this.columns = targetColumns;
+        this.synthetics = new HashMap<>();
+        PartitionName partitionName = table.isPartitioned()
+            ? PartitionName.fromIndexOrTemplate(indexName)
+            : null;
+        InputFactory inputFactory = new InputFactory(nodeCtx);
+        var referenceResolver = new RefResolver(symbolEval, partitionName, targetColumns, table);
+        Context<CollectExpression<IndexItem, Object>> ctxForRefs = inputFactory.ctxForRefs(
+            txnCtx,
+            referenceResolver
+        );
+        Function<ColumnIdent, Reference> getRef = table::getReference;
+        this.valueIndexers = new ArrayList<>(targetColumns.size());
+        for (var ref : targetColumns) {
+            this.valueIndexers.add(
+                ref.valueType().valueIndexer(table.ident(), ref, getFieldType, getRef)
+            );
+            addToValidate(table, ctxForRefs, ref);
+        }
+        this.tableConstraints = new ArrayList<>(table.checkConstraints().size());
+        for (var column : table.notNullColumns()) {
+            Reference ref = table.getReference(column);
+            assert ref != null : "Column in #notNullColumns must be available via table.getReference";
+            if (targetColumns.contains(ref)) {
+                columnConstraints.merge(ref.column(), new NotNull(column), MultiCheck::merge);
```

Comment:
is it ok that we don't add it to the `tableConstraints`?

Reply:
Should be - if it is in the `targetColumns` we'll process it with the values and the column constraint is good enough.

Reply:
ah, ok, thx! didn't get this